### ISSUE TITLE
Add command does not support json encoding

### ISSUE THREAD

**Issue by travisperson** [ipfs/go-ipfs#1121]
Currently the add command does not support JSON encoding.

This is due [to setting the output to nil](https://github.com/ipfs/go-ipfs/blob/master/core/commands/add.go#L139).

A fix to this is to detect the output format and skip this `PostRun` if the encoding is anything other than `Text`.

```go
v, found, err := req.Option(cmds.EncShort).String()
if err == nil && found && v != cmds.Text {
	log.Debugf("Encoding is not text, abort and let the marshaller handel it\n")
	return
}
```

This however introduced another issue. All output is returned as JSON. If the `previoususerProvidedEncoding` [is not set, we keep the JSON encoding](https://github.com/ipfs/go-ipfs/blob/master/commands/http/client.go#L114).

The reason we don't have a default encoding is due to using two different option keys, [in the client.go file we use](https://github.com/ipfs/go-ipfs/blob/master/commands/http/client.go#L39) `cmds.EncShort`, [while in the cli we use `encoding`](https://github.com/ipfs/go-ipfs/blob/db56c0f16a71b8cf36078a7233397de9d5b5f55a/cmd/ipfs/main.go#L239).

However, there is another issue, since the `add` command does not use a marshaler, [the check on line 240](https://github.com/ipfs/go-ipfs/blob/db56c0f16a71b8cf36078a7233397de9d5b5f55a/cmd/ipfs/main.go#L239) will fail and switches to text.

PostRun is used for progress, [and at some point the Marshaler was removed](https://github.com/ipfs/go-ipfs/commit/121dfb10b4b8d3a2fd4268943e65b30e5a8301a9#diff-bd0ca9f1e2d5c7ba128af01ec4c1131c).

I don't know the history of this file, so there might of been a good reason (not used?). Anyways without a Marshaler the original encoding will never get set back.

This is a start of a PR I want to open

```diff
diff --git a/cmd/ipfs/main.go b/cmd/ipfs/main.go
index 2ad90cf..411b833 100644
--- a/cmd/ipfs/main.go
+++ b/cmd/ipfs/main.go
@@ -236,11 +236,11 @@ func (i *cmdInvocation) Parse(ctx context.Context, args []string) error {
 
 	// if no encoding was specified by user, default to plaintext encoding
 	// (if command doesn't support plaintext, use JSON instead)
-	if !i.req.Option("encoding").Found() {
+	if !i.req.Option(cmds.EncShort).Found() {
 		if i.req.Command().Marshalers != nil && i.req.Command().Marshalers[cmds.Text] != nil {
-			i.req.SetOption("encoding", cmds.Text)
+			i.req.SetOption(cmds.EncShort, cmds.Text)
 		} else {
-			i.req.SetOption("encoding", cmds.JSON)
+			i.req.SetOption(cmds.EncShort, cmds.JSON)
 		}
 	}
 
diff --git a/core/commands/add.go b/core/commands/add.go
index 2b3297e..b2fc45f 100644
--- a/core/commands/add.go
+++ b/core/commands/add.go
@@ -127,7 +127,18 @@ remains to be implemented.
 			}
 		}()
 	},
+	Marshalers: cmds.MarshalerMap{
+		cmds.Text: func(res cmds.Response) (io.Reader, error) {
+			return nil, nil
+		},
+	},
 	PostRun: func(req cmds.Request, res cmds.Response) {
+		v, found, err := req.Option(cmds.EncShort).String()
+		if err == nil && found && v != cmds.Text {
+			log.Debugf("Encoding is not text, abort and let the marshaller handel it\n")
+			return
+		}
+
 		if res.Error() != nil {
 			return
 		}
```

I just want to get some feedback before making any commits.

**Comment by travisperson** [comment]
There is another issue with this that the `add` command does not support xml

```
Error: xml: unsupported type: <-chan interface {}
```

I have yet to tract down where the exact issue is though.

**Comment by travisperson** [comment]
Also pretty sure moving `encoding` to `cmds.EncShort`, may not be the exact solution. I believe other places send `encoding`, though we can probably just clean it up.

**Comment by whyrusleeping** [comment]
i wonder if we can remove the postrun thing entirely and get away with just getting fancy with the marshaler... I do a lot of hacky shit (tm) with the marshaler in core/commands/dht.go

**Comment by travisperson** [comment]
Ping

**Comment by jbenet** [comment]
what's next here?

**Comment by travisperson** [comment]
Basically we need to figure out what is the best approach to solve this issue. I played around with the idea @whyrusleeping mentioned but ran into concurrency issues and printing to stdout. I unfortunately accidently did a git reset and lost my changes, but the idea was to use a marshaller channel. 

This is the bit we talked about a while go: https://botbot.me/freenode/ipfs/2015-04-28/?msg=37740801&page=3

I'm not certain what the best approach is besides to not use the marshaller and simple correct (by applying the patch above or something similar) the issue of not resetting the encoding type.

**Comment by whyrusleeping** [comment]
heres what we can do, we can keep track of progress on the daemon side (within the Run function) and pass all the keys back as objects over the channel marshaler in an object that looks something like:

```go
type AddUpdate struct {
    Key string
    Progress float64
}
```

**Comment by chriscool** [comment]
@travisperson you had not committed your changes ?
Anyway most of the time using `git reset --keep` is the safest option to use when you want to reset. It will not remove uncommited changes in the working directory.

**Comment by travisperson** [comment]
Doesn't really matter too much. They were easy changes and I can rewrite
them if it's really the way we want to go. However,  I'm almost certain
it's not what we wanted to end up doing.

**Comment by whyrusleeping** [comment]
has this been fixed? @travisperson

**Comment by chriscool** [comment]
For this to move forward, it would be a good idea to write sharness tests even if some of them fail.
When some tests are not working it is always possible to use "test_expect_failure" instead of "test_expect_success" and to commit those tests anyway. It is a good way to document what is not working but should work.

**Comment by jbenet** [comment]
Agree with Chris wholeheartedly

**Comment by hsanjuan** [comment]
This is not fixed yet. Has the revamp of the commands lib made the fix easier maybe ?

**Comment by Stebalien** [comment]
No. See: https://github.com/ipfs/go-ipfs-cmds/issues/115.

**Comment by bqv** [comment]
This has been idle for two years, what's going on here?

**Comment by hsanjuan** [comment]
Hi @bqv . What's going on is that this is low priority because it is cosmetic and there are some workarounds, while the fix is not super simple and has been blocked on a number of refactors.
 
But I agree that the case of `/add` is the most annoying because curl-ing the API to add directories recursively is not feasible.

Also related #7050 .

**Comment by bqv** [comment]
Gotcha, ok
